### ISSUE TITLE
bug(lidar): unexpected serial resource check

### ISSUE THREAD

**Issue by evanchaoli** [concourse/concourse#4940]
## Bug Report

Our users report a check slowness issue, some resource checks took more than 5 minutes, even than 20 minutes.

I looked into the issue, and found the bug is most probably at: https://github.com/concourse/concourse/blob/master/atc/db/check.go#L253:

```go
func (c *check) AcquireTrackingLock(logger lager.Logger) (lock.Lock, bool, error) {
	return c.lockFactory.Acquire(
		logger,
		lock.NewResourceConfigCheckingLockID(c.ResourceConfigID()), // <<==HERE
	)
}
```
Our cluster enables Lidar but global resource. Here `check` use `c.ResourceConfigID()` as lock ID, which causes all checks of those resources having same configure to run in serial order.

See a real number from one of our clusters:

```sql
select count(distinct(pipeline_id)) from resources where resource_config_id=49950
count
------
65
```

There are 65 pipelines having resources with same configure, which means, the 65 pipelines will only run builds in serial.

When I look at those resources' configure, they all use variables like:

```yml
resources:
- name: mygit
  url: xxxxxx
  branch: master
  private-key: ((gitlab-key))
```

Though pipelines are from different teams, and their private keys are different, but as they use same variable name, Concourse would consider them as same. (Note: I didn't actually look through code to figure out if Concourse interpolate variables before save data to `resource_configs`, but I assume Concourse is not able to do so, because if user updates Key in Vault, Concourse has no way to detect the change and update `resource_configs`.)

To quickly solve the check slowness issue, I think we should use `resource_id` as lock id when global resource is not turned on.

A deeper thought: will using variable in resource configure kill `global resource`? Given people often copy pipeline YAML from samples, resource definitions like the above are normal. If multiple pipelines of different teams have the same resource definition, though their `gitlab-key` from above sample are actually different, Concourse may wrongly consider them as same. From the doc https://concourse-ci.org/global-resources.html, `global resource` considers only `type` and `source` configuration, however `team` and `pipeline` should be put into consideration as well, because a variable could be interpolated from either pipeline path or team path from Vault.

@vito I think this is a serious issue, WDYT? If you confirm this is a bug, I might raise a quick fix that would use `resource id` as lock id when `global resource` is disabled, and hopefully the fix can be included in 5.8.0.

## Steps to Reproduce

1. Enable `lidar` and disable `global resource` on a cluster.
2. Create 20 pipelines, each has a git resource with same `source`, the job should `get` the resource with `trigger: true`
3. Push a commit to the git repo
4. Watch the 20 pipelines, jobs will run in serial.

## Expected Results

The 20 pipelines will be trigger to run in parallel. 

## Actual Results

Jobs will run in serial.

## Version Info

* Concourse version: 5.7.2 (previous version should have same issue)
* Deployment type (BOSH/Docker/binary):
* Infrastructure/IaaS:
* Browser (if applicable):
* Did this used to work?

**Comment by vito** [comment]
> There are 65 pipelines having resources with same configure, which means, the 65 pipelines will only run builds in serial.

Can you clarify this? Only the *checks* should run serially, not the builds - and builds should only wait on checks if they're manually triggered. And manually triggered builds should only wait for the check to have happened *at all* after the build itself was created, so 65 triggered builds should all be good to go as soon as a single `check` finishes. Is that the case?

> If multiple pipelines of different teams have the same resource definition, though their `gitlab-key` from above sample are actually different, Concourse may wrongly consider them as same.

This would only be wrong if their version history actually differed based on the content of a var, which in the case of credentials is not typical. It *can* be the case if e.g. the credential determines a subset of content you can see, rather than just determining access, but this isn't typical behavior of `git` repos.

**Comment by evanchaoli** [comment]
> > There are 65 pipelines having resources with same configure, which means, the 65 pipelines will only run builds in serial.
> 
> Can you clarify this? Only the _checks_ should run serially, not the builds - and builds should only wait on checks if they're manually triggered. And manually triggered builds should only wait for the check to have happened _at all_ after the build itself was created, so 65 triggered builds should all be good to go as soon as a single `check` finishes. Is that the case?
> 

With Lidar, one thread creates check tuples in table `checks`, and the other thread reads check tuples from table `checks` then run checks, at this point, checks run in serial due to the lock. Without `check` returning new version, build will not run.

**Comment by vito** [comment]
Ah, I think I see what's happening; they're all queued up, but we only go over the queue periodically, and only one queue entry can check at a time because they all have the same lock.

So the builds aren't actually serial (I can actually see them running concurrently if the build runs long enough), but it's true that only one gets kicked off per LIDAR tick.

With global resources this isn't a big deal because one `check` entry succeeding counts as having `check`ed all of them.

@pivotal-jwinters Any thoughts on the proposed work-around? Maybe the lock should be on the *scope*?

**Comment by vito** [comment]
> > If multiple pipelines of different teams have the same resource definition, though their `gitlab-key` from above sample are actually different, Concourse may wrongly consider them as same.
> 
> This would only be wrong if their version history actually differed based on the content of a var, which in the case of credentials is not typical. It _can_ be the case if e.g. the credential determines a subset of content you can see, rather than just determining access, but this isn't typical behavior of `git` repos.

Actually, this felt a little off so I went and checked: version history is associated to the *fully evaluated* source configuration, so if you have two different credentials there won't be any issue - they will have distinct histories. Also different locks. So there is no risk of the "different subsets" issue I outlined.

**Comment by pivotal-jwinters** [comment]
@evanchaoli 

> (Note: I didn't actually look through code to figure out if Concourse interpolate variables before  save data to resource_configs, but I assume Concourse is not able to do so, because if user updates Key in Vault, Concourse has no way to detect the change and update resource_configs.)

That's actually exactly what we do :). We are constantly re-interpolating the source config and de-duplicating across the cluster every time we determine if new checks need to run. 

@vito 

The underlying problem is the ratio of `resource_config_scopes` to `resource_configs` because as you mentioned we schedule a check per `scope`, but they all check using the same container which is defined by the config.

You can actual use this query to determine your scope:config ratio:
`select resource_config_id, count(id) from resource_config_scopes group by resource_config_id;`

The real solution is https://github.com/concourse/concourse/issues/3424 so that we don't need to share check containers. In the interim we could create one check container per scope, however that would increase the number of check containers.

**Comment by pivotal-jwinters** [comment]
@vito The other option is we can just stop doing checks before a manually triggered build and we just run with whatever versions we have at the time ;).

**Comment by vito** [comment]
Seems like there are a few ways out here:

1. Do #3424 and stop re-using containers.
    * Last time we looked at this it got blocked on a lot of work. Is it time to re-assess? :eyes: 
1. Enable global resources by default (#4870).
    * We might be able to do this soon-ish; an answer for IAM auth is fast approaching with `var_sources`, primarily cluster-wide `var_sources`. Now we just need an IAM var source so that there's a pattern for resource types to adopt.
1. Hasten the LIDAR checking.
    * Maybe the flag is enough? Should we just bump the default if it's cheap enough?

> @vito The other option is we can just stop doing checks before a manually triggered build and we just run with whatever versions we have at the time ;).

This happens independently of manually triggered builds. Also that's kind of fundamental behavior of manually triggered builds in terms of what the user expects, so I don't think we'd ever change that.

**Comment by pivotal-jwinters** [comment]
> This happens independently of manually triggered builds. Also that's kind of fundamental behavior of manually triggered builds in terms of what the user expects, so I don't think we'd ever change that.

I would argue that there are actually two separate user needs here. One is `check all my resources`, and the other is `run this build`. And (to me anyways) it would be really nice to know exactly what versions a build is going to run with before clicking the `+`.

**Comment by evanchaoli** [comment]
Reducing check interval helps but I'm afraid that doesn't fundamentally solve the problem. From our database, I can see the longest queue check has been queued for more than 20 minutes.

In any case, when `global resource` is disabled, using an unique lock id for resources from different teams/pipelines sounds unreasonable, even if their configurable is exactly same, because each check still generates version history independently, and job trigger relies on its own pipeline's resource.

To reduce number of checks, why don't allow to skip check for some resources? For example, to expose build metadata (BUILD_NAME, etc) to tasks, we created a `buildinfo` resource type, per @vito's suggestion, that resource is only used with `put`, `check` doesn't do anything. The `buildinfo` resource is broadly used in a lot of pipelines, which generates a lot of checks on the resource, but those checks don't do anything at all. If Concourse supports something like:

```yaml
resource:
- name: build-metadata
  type: buildinfo
  skip_check: true <== this
```

Then I think that will a lot of checks and check containers.

**Comment by pivotal-jwinters** [comment]
@evanchaoli 

> In any case, when global resource is disabled, using an unique lock id for resources from different teams/pipelines sounds unreasonable, even if their configurable is exactly same, because each check still generates version history independently, and job trigger relies on its own pipeline resource's versions.

Versions are tied to `scopes`. So resources will share a version history if they have the same scope.

**Comment by pivotal-jwinters** [comment]
@evanchaoli 

> To reduce number of checks, why don't allow to skip check for some resources? 

This is interesting. There are two types of checks (manual and auto). We would probably want this to only apply to the auto checks, so that if a user wants to force a check they can.

**Comment by vito** [comment]
> To reduce number of checks, why don't allow to skip check for some resources? For example, to expose build metadata (BUILD_NAME, etc) to tasks, we created a `buildinfo` resource type, per @vito's suggestion, that resource is only used with `put`, `check` doesn't do anything. The `buildinfo` resource is broadly used in a lot of pipelines on our clusters, which generates a lot of checks on the resource, but those checks don't do anything at all. If Concourse supports something like:

I've thought about this before - I think if we did `check`s only for resources that are actually used in `get` steps we could achieve the same thing without adding a knob for it. @pivotal-jwinters I wonder if this would be a...moderately easy change to `LIDAR`? Especially now that we track job inputs in their own table? :eyes: 

> I would argue that there are actually two separate user needs here. One is `check all my resources`, and the other is `run this build`. And (to me anyways) it would be really nice to know exactly what versions a build is going to run with before clicking the `+`.

:+1: There's some prior art around this here: https://github.com/concourse/concourse/issues/413#issuecomment-233076618

**Comment by pivotal-jwinters** [comment]
@vito 

> I've thought about this before - I think if we did checks only for resources that are actually used in get steps we could achieve the same thing without adding a knob for it. @pivotal-jwinters I wonder if this would be a...moderately easy change to LIDAR? Especially now that we track job inputs in their own table? 👀

Ya! We would have to wait until we have `job_inputs` from 6.0, but otherwise seems like a win!

**Comment by evanchaoli** [comment]
> @evanchaoli
> 
> > In any case, when global resource is disabled, using an unique lock id for resources from different teams/pipelines sounds unreasonable, even if their configurable is exactly same, because each check still generates version history independently, and job trigger relies on its own pipeline resource's versions.
> 
> Versions are tied to `scopes`. So resources will share a version history if they have the same scope.

@pivotal-jwinters That is correct!!! I think that proves this bug, `check` should use `scope id` rather than `resource config id` as lock id. Do you agree?

**Comment by vito** [comment]
@evanchaoli That would fix this problem, but introduce another: if you associate the lock to the scope, you will end up with many concurrent `check`s running in the same container, because those are per-config (configs have many scopes). This could lead to breakage for resource types that assume their container is single-purpose - for example the `git` resource `check`s could end up stepping on each other's toes because there is local state on disk. It *might* be fine, but we can't guarantee that for every resource type ever written.

And if you make the containers per-scope, that will lead to a large increase in container count (65 more in your case), which could be an even worse problem at scale.

This is why we're leaning more towards ephemeral `check` containers - that way nothing ever runs concurrently in the same container, and we also don't have to worry about a constant inflation of container counts because they go away so quickly.

**Comment by evanchaoli** [comment]
@vito In that case, only thing we can do is to reduce `LidarCheckerInterval` to like 1 second.

**Comment by evanchaoli** [comment]
> And if you make the containers per-scope, that will lead to a large increase in container count (64 more in your case), which could be an even worse problem at scale.
> 

I think 64 more containers are OK to us, we are easy to add workers. And "global resource" will avoid such duplicated containers, if some users cannot afford such those extra containers, they may turn on "global resource", and looks like Concourse will make "global resource" as default in the future.

**Comment by vito** [comment]
The long-term fix for this will be to change `check` containers to be one-offs so we don't have to synchronize access to them. I'm making a new project for that [here](https://project.concourse-ci.org/projects/MDc6UHJvamVjdDM5Nzg3OTk=). With LIDAR this should be much easier to implement now than it was in our previous attempts, so I'm optimistic about getting this proper fix out shortly after 6.0. :+1:
